src/backend/distributed/utils/citus_clauses.c

```diff
@@ -29,6 +29,32 @@ static Expr * citus_evaluate_expr(Expr *expr, Oid result_type, int32 result_typm
 
 
 /*
+ * Whether the executor needs to reparse and try to execute this query.
+ */
+bool
+RequiresMasterEvaluation(Query *query)
+{
+	ListCell *targetEntryCell = NULL;
+
+	foreach(targetEntryCell, query->targetList)
```

Comment:
This adds yet another pass through most of the querytree. Doesn't necessarily have to be now, but can't we combine some of that?  Recursively iterating through pointered structures if far from cheap.